provider/throttle_utils.py

```diff
@@ -141,3 +152,600 @@ def change_throttle_group(self, image, group_id):
+        elif normal_read_iops:
+            mode = "randread"
+        elif normal_write_iops:
+            mode = "randwrite"
+        else:
+            mode = "randrw"
+
+        option += " --rw=%s --bs=%d --runtime=%d" % (mode, iops_size, runtime)
+
+        logging.debug(self._throttle["expected"])
+        logging.debug("fio_option:" + option)
+        self._fio_option = option
+
+    def build_image_option(self, image):
+        """
+        Build fio relevant info for image.
+
+        :param image:
+        :return: dict of image relevant data.
+        """
+
+        if image not in self._throttle["images"].keys():
+            self._throttle["images"].update({image: copy.deepcopy(
+                ThrottleTester.raw_image_data)})
+
+        name = get_drive_path(self._session, self._params, image)
+        image_data = self._throttle["images"][image]
+        image_data["name"] = name
+        image_data["fio_option"] = self._fio_option + " --filename=%s" % name
+        return image_data
+
+    def build_images_option(self):
+        """
+        Build fio relevant info for all images.
+
+        :return: dict of all images relevant data.
+        """
+
+        for image in self.images:
+            self.build_image_option(image)
+        return self._throttle["images"]
+
+    def attach_image(self, image):
+        """
+        Attach new image into throttle group.
+
+        :param image: image name.
+        """
+
+        self.images.append(image)
+
+    def detach_image(self, image):
+        """
+        Detach image from throttle group.
+
+        :param image: image name.
+        """
+
+        self.images.remove(image)
+
+
+class ThrottleGroupsTester(object):
+    """
+    This class mainly testing multi groups parallel or specified group
+    Example of usage:
+        ...
+        fio = generate_instance(params, vm, 'fio')
+        t1 = ThrottleTester("group1",["img1","img2"])
+        t1.set_fio(fio)
+        t1.build_default_option()
+        t1.build_images_option()
+        t2 = ThrottleTester("group1",["img1","img2"])
+        t2.set_fio(fio)
+        t2.build_default_option()
+        t2.build_images_option()
+        testers = ThrottleGroupsTester([t1,t2])
+        testers.start()
+    """
+
+    def __init__(self, testers):
+        self.testers = testers.copy()
+
+    @staticmethod
+    def proc_wrapper(func):
+        """Wrapper to log exception"""
+        try:
+            return func()
+        except Exception as e:
+            logging.exception(e)
+            raise
+
+    def start_group_test(self, group):
+        """
+        Start one group testing.
+
+        :param group:
+        :return:
+        """
+        for tester in self.testers:
+            if tester.group == group:
+                tester.start()
+                break
+
+    def start(self):
+        """
+        Start multi groups testing parallel.
+        """
+        num = len(self.testers)
+        pool = ThreadPool(num)
```

Comment:
there is an interface in vt: utils_misc.parallel, which can handle operations in parallel, just FYI.